frontend/src/metabase/query_builder/components/filters/pickers/SpecificDatePicker.tsx

```diff
@@ -0,0 +1,266 @@
+  }
+
+  let hasTime = false;
+  if (typeof hours === "number" && typeof minutes === "number") {
+    m.hours(hours);
+    m.minutes(minutes);
+    hasTime = true;
+  }
+
+  if (hasTime) {
+    return m.format(DATE_TIME_FORMAT);
+  } else {
+    return m.format(DATE_FORMAT);
+  }
+};
+
+type BetweenPickerProps = {
+  className?: string;
+  filter: Filter;
+  onFilterChange: (filter: any[]) => void;
+
+  hideTimeSelectors?: boolean;
+};
+
+export const BetweenPicker = ({
+  className,
+  filter: [op, field, startValue, endValue],
+  onFilterChange,
+  hideTimeSelectors,
+}: BetweenPickerProps) => (
+  <div className={className}>
+    <TimeContainer>
+      <div>
+        <SpecificDatePicker
+          value={startValue}
+          hideTimeSelectors={hideTimeSelectors}
+          onChange={value => onFilterChange([op, field, value, endValue])}
+        />
+      </div>
+      <div>
+        <SpecificDatePicker
+          value={endValue}
+          hideTimeSelectors={hideTimeSelectors}
+          onClear={() =>
+            onFilterChange([
+              op,
+              field,
+              setTimeComponent(startValue),
+              setTimeComponent(endValue),
+            ])
+          }
+          onChange={value => onFilterChange([op, field, startValue, value])}
+        />
+      </div>
+    </TimeContainer>
+    <div className="Calendar--noContext">
+      <Calendar
+        initial={startValue ? moment(startValue) : moment()}
+        selected={startValue && moment(startValue)}
+        selectedEnd={endValue && moment(endValue)}
+        onChange={(startValue, endValue) =>
+          onFilterChange([op, field, startValue, endValue])
+        }
+      />
+    </div>
+  </div>
+);
+
+type SingleDatePickerProps = {
+  className?: string;
+  filter: Filter;
+  selectAll?: SelectAll;
+  onFilterChange: (filter: any[]) => void;
+
+  hideTimeSelectors?: boolean;
+};
+
+export const SingleDatePicker = ({
+  className,
+  filter: [op, field, value],
+  onFilterChange,
+  hideTimeSelectors,
+  selectAll,
+}: SingleDatePickerProps) => (
+  <SpecificDatePicker
+    className={className}
+    value={value}
+    selectAll={selectAll}
+    onChange={value => onFilterChange([op, field, value])}
+    onClear={() => onFilterChange([op, field, setTimeComponent(value)])}
+    hideTimeSelectors={hideTimeSelectors}
+    calendar
+  />
+);
+
+export const BeforePicker = (props: SingleDatePickerProps) => {
+  return <SingleDatePicker {...props} selectAll="before" />;
+};
+
+export const AfterPicker = (props: SingleDatePickerProps) => {
+  return <SingleDatePicker {...props} selectAll="after" />;
+};
+
+type Props = {
+  className?: string;
+  calendar?: boolean;
+  selectAll?: SelectAll;
+
+  hideTimeSelectors?: boolean;
+  value: string;
+  onChange: (startValue: string | null, endValue?: string) => void;
+  onClear?: () => void;
+};
+
+type State = {
+  showCalendar: boolean;
+};
+
+export default class SpecificDatePicker extends Component<Props, State> {
```

Comment:
Can we convert it to a functional component? We try to avoid creating new class components

Reply:
Done